### PR TITLE
indexer: extend to improve archive support

### PR DESCRIPTION
**Pull request by p4u**
- add startTime and endTime to indexer, this way we do not need the block time estimation.

- add the chainID into the indexer database, so we can identify the origin chain

- add fromArchive flag to inform the user about the process being added by the archive

### CODE REVIEW

File: vochain/indexer/archive.go [hunk 1]
```diff
@@ -61,17 +61,33 @@ func (idx *Indexer) ImportArchive(archive []*ArchiveProcess) ([]*ArchiveProcess,
...
+			}
+		}
+		endDate := p.ProcessInfo.EndDate
+		if endDate.IsZero() {
+			// Calculate endDate equal to startDate plus defaultBlockTime*p.ProcessInfo.BlockCount
+			endDate = startDate.Add(defaultBlockTime * time.Duration(p.ProcessInfo.BlockCount))
 		}
+
 		// Create and store process in the indexer database
 		procParams := indexerdb.CreateProcessParams{
```
Comment (altergui) [inline vochain/indexer/archive.go:82]: why are we not storing the ManuallyEnded bool anywhere?

Comment (altergui) [reply vochain/indexer/archive.go:82]: ah, now i see, ManuallyEnded is calculated based on `pi.EndBlock < pi.StartBlock+pi.BlockCount` , sorry for the noise

File: vochain/indexer/archive.go [hunk 2]
```diff
@@ -61,17 +61,33 @@ func (idx *Indexer) ImportArchive(archive []*ArchiveProcess) ([]*ArchiveProcess,
...
 			ID:                nonNullBytes(p.ProcessInfo.ID),
 			EntityID:          nonNullBytes(p.ProcessInfo.EntityID),
-			StartBlock:        int64(height),
-			EndBlock:          int64(height + 1),
-			BlockCount:        int64(1),
+			StartBlock:        int64(p.ProcessInfo.StartBlock),
+			StartDate:         startDate,
+			EndBlock:          int64(p.ProcessInfo.EndBlock),
+			EndDate:           endDate,
```
Comment (altergui) [inline vochain/indexer/archive.go:88]: to me it looks somewhat confusing, prone to error (or redundant?), that we're storing both Block and Date... blocks already have a very definite date

Comment (p4u) [reply vochain/indexer/archive.go:88]: The whole point for this PR is to be able to store processes from previous (dead) chains. This is the `archive`.
So we don't have a way to estimate the start/endblock on those processes.

File: vochain/indexer/archive.go [hunk 3]
```diff
@@ -85,13 +101,16 @@ func (idx *Indexer) ImportArchive(archive []*ArchiveProcess) ([]*ArchiveProcess,
...
 			Metadata:          p.ProcessInfo.Metadata,
 			ResultsVotes:      indexertypes.EncodeJSON(p.Results.Votes),
 			VoteCount:         int64(p.ProcessInfo.VoteCount),
+			ChainID:           p.ChainID,
+			FromArchive:       true,
 		}
+
 		if _, err := queries.CreateProcess(context.TODO(), procParams); err != nil {
 			return nil, fmt.Errorf("create archive process: %w", err)
 		}
```
Comment (altergui) [inline vochain/indexer/archive.go:115]: you could seize the opportunity to fix the typo in the next func name `StartArchiveRetrival` -> `StartArchiveRetrieval`

File: vochain/vochaininfo/vochaininfo.go [hunk 4]
```diff
@@ -14,6 +14,10 @@ import (
 	"go.vocdoni.io/dvote/vochain/state"
 )
 
+const (
+	defaultBlockTime = 10000 // 10 seconds
```
Comment (altergui) [inline vochain/vochaininfo/vochaininfo.go]: why is this different from the other defaultBlockTime (12s)?

File: vochain/vochaininfo/vochaininfo.go [hunk 5]
```diff
@@ -68,14 +72,17 @@ func (vi *VochainInfo) updateCounters() {
 	mempoolSize.Set(uint64(vi.vnode.MempoolSize()))
 }
 
-// Height returns the current number of blocks of the blockchain
+// Height returns the current number of blocks of the blockchain.
```
Comment (altergui) [inline vochain/vochaininfo/vochaininfo.go:72]: ???? :rofl: if you really want to lint this, make it on a separate commit linting the thousand comments with missing dot

File: vochain/vochaininfo/vochaininfo.go [hunk 6]
```diff
@@ -68,14 +72,17 @@ func (vi *VochainInfo) updateCounters() {
...
 func (vi *VochainInfo) Height() uint64 {
 	return height.Get()
 }
 
 // BlockTimes returns the average block time in milliseconds for 1, 10, 60, 360 and 1440 minutes.
-// Value 0 means there is not yet an average
+// Value 0 means there is not yet an average.
 func (vi *VochainInfo) BlockTimes() *[5]uint64 {
+	if vi.vnode.IsSynchronizing() {
+		return &[5]uint64{defaultBlockTime, 0, 0, 0, 0}
```
Comment (altergui) [inline vochain/vochaininfo/vochaininfo.go]: i'd avoid changing BlockTimes, which is called in many places. 
in fact there's already a fallback in HeightTime and EstimateBlockHeight, so i pushed a commit with a better fix

**Review by altergui — APPROVED** [review]
in general LGTM so approving with the usual faith, although i'm a bit lost on why are we storing Date and Block, and ImportArchive is duplicating the code for calculating startDate and endDate, instead of using one of the already existing func HeightTime

File: vochain/vochaininfo/vochaininfo.go (hunk 4 above)
Comment (p4u) [reply vochain/vochaininfo/vochaininfo.go]: Actually we should unify this into type types package

**Comment by p4u** [comment]
> in general LGTM so approving with the usual faith, although i'm a bit lost on why are we storing Date and Block, and ImportArchive is duplicating the code for calculating startDate and endDate, instead of using one of the already existing func HeightTime

We don't have the height time from a Chain which is not the existing one, @altergui, because we don't have the blocks neither the headers. What is confusing you about this?

I.e in this scenario, our current chain could be at block 1000 but the startBlock of the archive process is 500000. We don't want to return startDate=future while the process is actually finished and was closed months ago. We need to have the exact startDate at which the process was created on the previous chain.

File: vochain/vochaininfo/vochaininfo.go (hunk 6 above)
Comment (p4u) [reply vochain/vochaininfo/vochaininfo.go]: Umm, but if the chain is synchronizing the blocktimes will have completely absurd values (such as 0.005). So having a fallback for such scenario that returns a kind of correct value, makes sense IMHO.
